trio/_wait_for_object.py

```diff
@@ -1,62 +1,200 @@
+        self._handle_map = {}
+        self._size_sorted_wait_groups = SortedKeyList(key=len)
+        self.lock = lock
+
+    def __contains__(self, item):
+        return item in self._handle_map
+
+    def add(self, handle, callback):
+        # Shortcut if we are already waiting on this handle
+        if handle in self._handle_map:
+            self._handle_map[handle][0].add(callback)
+            return
+
+        try:
+            wait_group = self._size_sorted_wait_groups.pop()
+        except IndexError:
+            # pool is empty or every group is full
+            wait_group = spawn_wait_group()
+        else:
+            kernel32.SetEvent(wait_group[0])
+
+        self._handle_map[handle] = ({callback}, wait_group)
+
+        with self.mutating(wait_group):
+            wait_group.append(handle)
+
+    def remove(self, handle, callback):
+        callbacks, wait_group = self._handle_map[handle]
+
+        callbacks.remove(callback)
+
+        if callbacks:
+            # no cleanup or thread interaction needed
+            return
+
+        del self._handle_map[handle]
+
+        with self.mutating(wait_group):
+            wait_group.remove(handle)
+
+        if len(wait_group) == 1:
+            # need a wake to make sure this thread exits promptly
+            kernel32.SetEvent(wait_group[0])
+
+    def execute_callbacks(self, handle):
+        # also discards our internal reference to the handle
+        for callback in self._handle_map.pop(handle)[0]:
+            callback()
+
+    @contextlib.contextmanager
+    def mutating(self, wait_group):
+        # SortedKeyList can't cope with mutating keys, so
+        # remove and re-add to maintain sort order
+        self._size_sorted_wait_groups.discard(wait_group)
+        try:
+            yield
+        finally:
+            # if SortedKeyList has many equal values, remove() performance degrades
+            # from O(log(n)) to O(n), so we don't keep full or empty groups inside
+            if 1 < len(wait_group) < MAXIMUM_WAIT_OBJECTS:
+                self._size_sorted_wait_groups.add(wait_group)
+
+
+WAIT_POOL = WaitPool(threading.Lock())
+
+
+def spawn_wait_group():
+    """Only to be used within WaitPool"""
+    # maintain a wake handle at index 0
+    wait_group = [kernel32.CreateEventA(ffi.NULL, True, False, ffi.NULL)]
+    trio_token = _core.current_trio_token()
+
+    def drain_as_completed():
+        # This with block prevents a race with the spawning thread on init
+        with WAIT_POOL.lock:
+            n_handles = len(wait_group) - 1
+        while n_handles:  # quit thread when only wake handle remains
+            signaled_handle = WaitForMultipleObjects_sync(*wait_group)
+            with WAIT_POOL.lock:
+                if signaled_handle is wait_group[0]:
+                    # Reset is OK here because only this thread waits on the event
+                    kernel32.ResetEvent(wait_group[0])
+                elif signaled_handle in WAIT_POOL:
```

Comment:
For what it's worth, codecov complains that we never evaluate this condition to False

Reply:
Yeah, I think it needs a pragma: no branch, unless I can figure out how to trigger the race mentioned in the comment in the next block